README.md
# Overview

This smart contract enables voting on proposals using ERC20 tokens.

## Tooling

Tests and tools are implemented using the
[chainlib-eth](https://pypi.org/project/chainlib-eth) python package.

To learn more about the individual CLI tools mentioned in this document,
please refer to their individual man pages. The man pages should be
installed as part of the python packages.

## Interoperability

The `evm-tokenvote` contract implements the `cic-contracts:TokenVote`
interface.

## Publishing the contract

In python, the smart contract can be published using the `constructor`
method of the `evm_tokenvote.Voter` class.

Also, constructor bytecode can be generated on the command-line using
`chainlib-gen`. Thereafter, `eth-gen` can be used to publish the
contract through an rpc endpoint:

    # Make sure evm_tokenvote is in your PYTHONPATH
    $ chainlib-gen evm_tokenvote create --token_address <token_address> > data.bin
    $ eth-gas -p <rpc_url> -y <json_keyfile> [ --passphrase-file <file_with_keyfile_passphrase> ] --data data.bin -i <evm:chain:chain_id:common_name> -s -w 0

The `token_address` argument **MUST** be an existing and valid `ERC20`
contract.

# Proposal

Proposals are created with a 32-byte description.

How the description should be interpreted is up to the client
application.

A proposal may be a binary (yes or no) vote, or define a number of
different options to choose from.

## Parameters

The arguments required to create a proposal depends on the type of
proposal to create.

### Deadline

The deadline must be defined for all proposals.

It is defined as a period of number of blocks during which votes may be
made.

If a vote has been completed before the deadline is over, the result
will be available immediately.

The contract does not check the sanity of the deadline value. For
example, setting the deadline to `1` block wait will be useless as
voting will expire immediately.

## Target vote

The target vote must be defined for all proposals.

This defines the minimum participation in the vote. It is defined as
parts-per-million of the total supply of tokens.

For example. a value of `500000` will require 50% of all tokens in the
vote. A value of `1000000` will require the full supply.

## Options

A proposal may define one or more options for voters to choose between.

Options are defined as 32-byte hexadecimal values. As with the
description, it is up to the client application to decide how to
interpret the values.

## Creating a proposal

To create a proposal without options:

    # solidity:
    function propose(bytes32 _description, uint256 _blockWait, uint24 _targetVotePpm) public returns (uint256);

    # chainlib-python:
    def propose(self, contract_address, sender_address, description, block_deadline, tx_format=TxFormat.JSONRPC, id_generator=None)

    # eth-encode CLI:
    $ eth-encode --mode tx --signature propose -e <voter_contract_address> -y <keyfile_json> a:<token_address> u:<blocks_until_deadline> u:<target_vote_ppm>

To create a proposal with options:

    solidity:
    function proposeMulti(bytes32 _description, uint256 _blockWait, uint24 _targetVotePpm) public returns (uint256);

    chainlib-python:
    def propose(self, contract_address, sender_address, description, block_deadline, options=[<options_hex>, ...])

(Unfortunately, `eth-encode` does not currently support dynamic array
arguments.)

# Voting

Votes are defined in magnitudes of ERC20 tokens.

A vote will transfer ERC20 tokens to the custody of the smart contract
for the duration of the voting.

The smart contract uses the `transferFrom` method to transfer tokens. It
is the caller’s responsibility to make the necessary token allowance
(using `approve`). Votes with insufficient allowance will fail.

## Proposal context

Votes are always cast on the oldest proposal the has not been completed.

## Options

If multiple options exist, the token holder may freely distribute votes
between the options.

## Cancellation votes

In both proposals with and without options, votes can be cast to cancel
the proposal.

For proposals with no or one option, the *cancel* amounts to a vote
against the proposal.

For proposals with two or more options, the *cancel* amounts to a vote
against proposal and its all options.

## How to vote

In each case, make sure the necessary *allowance* has been successfully
executed.

For proposals without options, a simplified method can be called:

    # solidity:
    function vote(uint256 _value) public returns (bool);


    # chainlib-python:
    def vote(self, contract_address, sender_address, value)


    # eth-encode CLI:
    $ eth-encode --mode tx --signature vote -e <voter_contract_address> -y <keyfile_json> u:<value>

For proposal with options, the call is slightly more verbose:

    # solidity:
    function voteOption(uint256 _optionIndex, uint256 _value) public returns (bool);


    # chainlib-python:
    def vote(self, contract_address, sender_address, option=<option_index>, value)


    # eth-encode CLI:
    $ eth-encode --mode tx --signature voteOption -e <voter_contract_address> -y <keyfile_json> u:<option_index> u:<value>

To cast votes for cancellation, the call will be:

    # solidity:
    function voteCancel(uint256 _value) public returns (bool);


    # chainlib-python:
    def vote_cancel(self, contract_address, sender_address, value)


    # eth-encode CLI:
    $ eth-encode --mode tx --signature voteCancel -e <voter_contract_address> -y <keyfile_json> u:<value>

# Results

A proposal vote is completed if either of the following are true:

- The deadline has been reached

- Proposal contains zero or one options, and target participation has
  been reached.

- Cancellation votes have the majority.

## Finalizing the proposal

Once a proposal vote has been completed, the proposal must be explicitly
finalized.

Finalization analyzes the results of the vote, and marks the proposal
state accordingly.

It also moves the proposal cursor to activate the next proposal in the
queue.

Finally, it releases the ERC20 tokens used for the vote.

Finalization is performed using the `finalize()` contract method. It
will fail if used before the proposal vote has been *completed*.

### Enhanced results

The optional method `scan(uint256 _proposalIndex, uint256 _count` can be
called on a completed proposal to further analyze the results of the
vote.

In the current state of the contract, it will iterate the options of the
proposal, and mark the state as `TIED` if two or more options have the
same amount of votes.

This method may be called any time after proposal has been completed
(even before `finalize()`. The proposal is identified by the
`_proposalIndex` parameter, where the index is the order of addition of
the proposal.

The `_count` parameter limits the amount of options that will be
scanned. A consecutive call will start at the option where the previous
left off.

## Recovering tokens

Before a new vote can take place, the ERC20 tokens used in previous
voting must be withdrawn.

Withdrawal is performed using the `withdraw()` contract method. It will
fail if used before the proposal vote has been *finalized*.

# Proposal states

| label          | bit value | description                                                                                                                                                          |
|----------------|-----------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| INIT           | 1         | Proposal has been initiated. It is used to disambiguate a proposal struct that has not yet been added, due to the ambiguity of the default struct value in solidity. |
| `FINAL`        | 2         | `finalize()` has been successfully called.                                                                                                                           |
| `SCANNED`      | 4         | `scan(...)` has been successfully called for all available options.                                                                                                  |
| `INSUFFICIENT` | 8         | Voting participation did not been the required target vote before the deadline.                                                                                      |
| `TIED`         | 16        | Proposal contains two or more options, and two or more options have received the same amount of votes before the deadline.                                           |
| `SUPPLYCHANGE` | 32        | The token supply changed between when the proposal was added and voting ended. If supply is protected, this has also `CANCELLED` the vote.                           |
| `IMMEDIATE`    | 64        | Voting was completed before the deadline.                                                                                                                            |
| `CANCELLED`    | 128       | Interpretation depends on context. See below.                                                                                                                        |

Interpreting `CANCELLED`:

- With `SUPPLYCHANGE`, this means that vote has been invalidated and
  should be discarded.

- If proposal has options, it should be interpreted as that the
  existence of the proposal itself has been *rejected*.

- With no or one option, it should be interpreted as the proposal has
  been *defeated*.


python/CHANGELOG
- 0.0.3
	* Introduce block wait limit
	* Add internal state change proposal mode
	* Implement change of block wait limit through (internal) proposal
- 0.0.2
	* Separate option addition from proposal creation 
- 0.0.1
	* Create proposals, with our without options
	* Vote with ERC20 tokens


python/tests/test_token.py
# standard imports
import unittest
import logging
import os
from chainlib.eth.nonce import RPCNonceOracle
from chainlib.eth.tx import receipt
from chainlib.eth.block import block_latest
from hexathon import same as same_hex
from eth_erc20 import ERC20
from giftable_erc20_token import GiftableToken

# local imports
from evm_tokenvote.unittest import TestEvmVoteProposal
from evm_tokenvote.unittest.base import hash_of_foo
from evm_tokenvote import Voter
from evm_tokenvote import ProposalState


logging.basicConfig(level=logging.DEBUG)
logg = logging.getLogger()


class TestVoteToken(TestEvmVoteProposal):

    def test_withdraw(self):
        c = Voter(self.chain_spec)
        o = c.get_proposal(self.voter_address, 0, sender_address=self.accounts[0])
        r = self.rpc.do(o)
        proposal = c.parse_proposal(r)

        half_supply = self.initial_supply / 2
        nonce_oracle = RPCNonceOracle(self.accounts[0], conn=self.conn)
        c = ERC20(self.chain_spec, signer=self.signer, nonce_oracle=nonce_oracle)
        (tx_hash, o) = c.transfer(self.address, self.accounts[0], self.alice, half_supply)
        self.rpc.do(o)

        nonce_oracle = RPCNonceOracle(self.alice, conn=self.conn)
        c = ERC20(self.chain_spec, signer=self.signer, nonce_oracle=nonce_oracle)
        (tx_hash, o) = c.approve(self.address, self.alice, self.voter_address, half_supply)
        self.rpc.do(o)

        c = Voter(self.chain_spec, signer=self.signer, nonce_oracle=nonce_oracle)
        (tx_hash, o) = c.vote(self.voter_address, self.alice, half_supply)
        self.rpc.do(o)

        c = ERC20(self.chain_spec)
        o = c.balance_of(self.voter_address, self.alice, sender_address=self.accounts[0])
        r = self.rpc.do(o)
        balance = int(r, 16)
        self.assertEqual(balance, half_supply)

        o = c.balance_of(self.address, self.alice, sender_address=self.accounts[0])
        r = self.rpc.do(o)
        balance = int(r, 16)
        self.assertEqual(balance, 0)

        nonce_oracle = RPCNonceOracle(self.alice, conn=self.conn)
        c = Voter(self.chain_spec, signer=self.signer, nonce_oracle=nonce_oracle)
        (tx_hash, o) = c.withdraw(self.voter_address, self.alice)
        self.rpc.do(o)
        o = receipt(tx_hash)
        r = self.rpc.do(o)
        self.assertEqual(r['status'], 0)

        c = ERC20(self.chain_spec, signer=self.signer, nonce_oracle=nonce_oracle)
        o = c.balance_of(self.voter_address, self.alice, sender_address=self.accounts[0])
        r = self.rpc.do(o)
        balance = int(r, 16)
        self.assertEqual(balance, half_supply)

        o = c.balance_of(self.address, self.alice, sender_address=self.accounts[0])
        r = self.rpc.do(o)
        balance = int(r, 16)
        self.assertEqual(balance, 0)

        c = Voter(self.chain_spec)
        o = c.get_proposal(self.voter_address, 0, sender_address=self.alice)
        r = self.rpc.do(o)
        proposal = c.parse_proposal(r)
        self.assertEqual(proposal.total, half_supply)

        o = block_latest()
        now_block_height = self.rpc.do(o)
        need_blocks = self.proposal_block_height + 100 - now_block_height + 1
        self.backend.mine_blocks(need_blocks)

        # after deadline withdraw is locked
        nonce_oracle = RPCNonceOracle(self.alice, conn=self.conn)
        c = Voter(self.chain_spec, signer=self.signer, nonce_oracle=nonce_oracle)
        (tx_hash, o) = c.withdraw(self.voter_address, self.alice)
        self.rpc.do(o)
        o = receipt(tx_hash)
        r = self.rpc.do(o)
        self.assertEqual(r['status'], 0)

        o = c.get_proposal(self.voter_address, 0, sender_address=self.accounts[0])
        r = self.rpc.do(o)
        proposal = c.parse_proposal(r)
        self.assertEqual(proposal.total, half_supply)

        c = ERC20(self.chain_spec)
        o = c.balance_of(self.voter_address, self.alice, sender_address=self.accounts[0])
        r = self.rpc.do(o)
        balance = int(r, 16)
        self.assertEqual(balance, half_supply)

        o = c.balance_of(self.address, self.alice, sender_address=self.accounts[0])
        r = self.rpc.do(o)
        balance = int(r, 16)
        self.assertEqual(balance, 0)

        # scan + finalize unlocks tokens
        nonce_oracle = RPCNonceOracle(self.trent, conn=self.conn)
        c = Voter(self.chain_spec, signer=self.signer, nonce_oracle=nonce_oracle)
        (tx_hash, o) = c.scan(self.voter_address, self.trent, 0, 0)
        self.rpc.do(o)

        (tx_hash, o) = c.finalize_vote(self.voter_address, self.trent)
        self.rpc.do(o)
        o = receipt(tx_hash)
        r = self.rpc.do(o)
        self.assertEqual(r['status'], 1)

        nonce_oracle = RPCNonceOracle(self.alice, conn=self.conn)
        c = Voter(self.chain_spec, signer=self.signer, nonce_oracle=nonce_oracle)
        o = c.get_proposal(self.voter_address, 0, sender_address=self.accounts[0])
        r = self.rpc.do(o)
        proposal = c.parse_proposal(r)
        self.assertEqual(proposal.total, half_supply)

        (tx_hash, o) = c.withdraw(self.voter_address, self.alice)
        self.rpc.do(o)
        o = receipt(tx_hash)
        r = self.rpc.do(o)
        self.assertEqual(r['status'], 1)

        c = ERC20(self.chain_spec)
        o = c.balance_of(self.address, self.alice, sender_address=self.accounts[0])
        r = self.rpc.do(o)
        balance = int(r, 16)
        self.assertEqual(balance, half_supply)

        o = c.balance_of(self.voter_address, self.alice, sender_address=self.accounts[0])
        r = self.rpc.do(o)
        balance = int(r, 16)
        self.assertEqual(balance, 0)



if __name__ == '__main__':
    unittest.main()


python/tests/test_cancel.py
# standard imports
import unittest
import logging
import os
from chainlib.eth.nonce import RPCNonceOracle
from chainlib.eth.tx import receipt
from chainlib.eth.block import block_latest
from hexathon import same as same_hex
from eth_erc20 import ERC20
from giftable_erc20_token import GiftableToken

# local imports
from evm_tokenvote.unittest import TestEvmVoteProposal
from evm_tokenvote.unittest.base import hash_of_foo
from evm_tokenvote import Voter
from evm_tokenvote import ProposalState


logging.basicConfig(level=logging.DEBUG)
logg = logging.getLogger()

class TestVoteBase(TestEvmVoteProposal):

    def test_cancel(self):
        half_supply = int(self.initial_supply / 2)
        nonce_oracle = RPCNonceOracle(self.accounts[0], conn=self.conn)
        c = ERC20(self.chain_spec, signer=self.signer, nonce_oracle=nonce_oracle)
        (tx_hash, o) = c.transfer(self.address, self.accounts[0], self.alice, half_supply)
        self.rpc.do(o)

        nonce_oracle = RPCNonceOracle(self.alice, conn=self.conn)
        c = ERC20(self.chain_spec, signer=self.signer, nonce_oracle=nonce_oracle)
        (tx_hash, o) = c.approve(self.address, self.alice, self.voter_address, half_supply)
        self.rpc.do(o)

        c = Voter(self.chain_spec, signer=self.signer, nonce_oracle=nonce_oracle)
        (tx_hash, o) = c.vote_cancel(self.voter_address, self.alice, half_supply - 1)
        self.rpc.do(o)
        o = receipt(tx_hash)
        r = self.rpc.do(o)
        self.assertEqual(r['status'], 1)

        nonce_oracle = RPCNonceOracle(self.trent, conn=self.conn)
        c = Voter(self.chain_spec, signer=self.signer, nonce_oracle=nonce_oracle)
        (tx_hash, o) = c.scan(self.voter_address, self.trent, 0, 0)
        self.rpc.do(o)
        o = receipt(tx_hash)
        r = self.rpc.do(o)
        self.assertEqual(r['status'], 0)

        nonce_oracle = RPCNonceOracle(self.alice, conn=self.conn)
        c = Voter(self.chain_spec, signer=self.signer, nonce_oracle=nonce_oracle)
        (tx_hash, o) = c.vote_cancel(self.voter_address, self.alice, 1)
        self.rpc.do(o)
        o = receipt(tx_hash)
        r = self.rpc.do(o)
        self.assertEqual(r['status'], 1)

        nonce_oracle = RPCNonceOracle(self.trent, conn=self.conn)
        c = Voter(self.chain_spec, signer=self.signer, nonce_oracle=nonce_oracle)
        (tx_hash, o) = c.scan(self.voter_address, self.trent, 0, 0)
        self.rpc.do(o)
        o = receipt(tx_hash)
        r = self.rpc.do(o)
        self.assertEqual(r['status'], 1)

        o = c.get_proposal(self.voter_address, 0, sender_address=self.accounts[0])
        r = self.rpc.do(o)
        proposal = c.parse_proposal(r)
        self.assertEqual(proposal.state & ProposalState.SCANNED, ProposalState.SCANNED)
        self.assertEqual(proposal.state & ProposalState.IMMEDIATE, ProposalState.IMMEDIATE)
        self.assertEqual(proposal.state & ProposalState.CANCELLED, ProposalState.CANCELLED)


if __name__ == '__main__':
    unittest.main()


python/evm_tokenvote/voter.py
# standard imports
import logging
import os
import enum

# external imports
from chainlib.eth.constant import ZERO_ADDRESS
from chainlib.eth.contract import (
    ABIContractEncoder,
    ABIContractDecoder,
    ABIContractType,
    abi_decode_single,
)
from chainlib.eth.jsonrpc import to_blockheight_param
from chainlib.eth.error import RequestMismatchException
from chainlib.eth.tx import (
    TxFactory,
    TxFormat,
)
from chainlib.jsonrpc import JSONRPCRequest
from chainlib.block import BlockSpec
from hexathon import (
    add_0x,
    strip_0x,
)
from chainlib.eth.cli.encode import CLIEncoder

# local imports
from evm_tokenvote.data import data_dir

logg = logging.getLogger()


class ProposalState(enum.IntEnum):
    INIT = 1
    FINAL = 2
    SCANNED = 4
    INSUFFICIENT = 8
    TIED = 16
    SUPPLYCHANGE = 32
    IMMEDIATE = 64
    CANCELLED = 128


class Proposal:
   
    def __init__(self, description_digest, *args, **kwargs):
        self.description_digest = description_digest
        self.supply = kwargs.get('supply')
        self.total = kwargs.get('total')
        self.block_deadline = kwargs.get('block_deadline')
        self.target_vote_ppm = kwargs.get('target_vote_ppm')
        self.cancel_votes = kwargs.get('cancel_votes')
        self.proposer = kwargs.get('proposer')
        self.state = kwargs.get('state')
        self.serial = kwargs.get('serial')


    def __str__(self):
        return "proposal description {} total {} supply {}".format(self.description_digest, self.total, self.supply)


class Voter(TxFactory):

    __abi = None
    __bytecode = None

    def constructor(self, sender_address, token_address, protect_supply=False, voter_registry=None, proposer_registry=None, tx_format=TxFormat.JSONRPC, version=None):
        code = self.cargs(token_address, protect_supply=protect_supply, voter_registry=voter_registry, proposer_registry=proposer_registry, version=version)
        tx = self.template(sender_address, None, use_nonce=True)
        tx = self.set_code(tx, code)
        return self.finalize(tx, tx_format)


    @staticmethod
    def cargs(token_address, protect_supply=False, voter_registry=None, proposer_registry=None, version=None):
        if voter_registry == None:
            voter_registry = ZERO_ADDRESS
        if proposer_registry == None:
            proposer_registry = ZERO_ADDRESS
        if token_address == None:
            raise ValueError("token address cannot be zero address")
        code = Voter.bytecode(version=version)
        enc = ABIContractEncoder()
        enc.address(token_address)
        enc.bool(protect_supply)
        enc.address(voter_registry)
        enc.address(proposer_registry)
        args = enc.get()
        code += args
        logg.debug('constructor code: ' + args)
        return code


    @staticmethod
    def gas(code=None):
        return 4000000


    @staticmethod
    def abi():
        if Voter.__abi == None:
            f = open(os.path.join(data_dir, 'Voter.json'), 'r')
            Voter.__abi = json.load(f)
            f.close()
        return Voter.__abi


    @staticmethod
    def bytecode(version=None):
        if Voter.__bytecode == None:
            f = open(os.path.join(data_dir, 'Voter.bin'))
            Voter.__bytecode = f.read()
            f.close()
        return Voter.__bytecode


    def propose(self, contract_address, sender_address, description, block_deadline, target_vote_ppm=500000, tx_format=TxFormat.JSONRPC, id_generator=None):
        enc = ABIContractEncoder()
        enc.method('propose')
        enc.typ(ABIContractType.BYTES32)
        enc.typ(ABIContractType.UINT256)
        enc.typ_literal('uint24')
        enc.bytes32(description)
        enc.uint256(block_deadline)
        enc.uintn(target_vote_ppm, 24)
        data = add_0x(enc.get())
        tx = self.template(sender_address, contract_address, use_nonce=True)
        tx = self.set_code(tx, data)
        tx = self.finalize(tx, tx_format, id_generator=id_generator)
        return tx

    
    def propose_blockwait(self, contract_address, sender_address, blockwait, block_deadline, target_vote_ppm=500000, tx_format=TxFormat.JSONRPC, id_generator=None):
        enc = ABIContractEncoder()
        enc.method('proposeInternal')
        enc.typ(ABIContractType.BYTES32)
        enc.typ(ABIContractType.BYTES32)
        enc.typ(ABIContractType.UINT256)
        enc.typ_literal('uint24')
        enc.bytes32('67ca084db32598c571e2ad2dc8b95679c3fa14c63213935dfd8f0a158ff65c57')
        blockwait_bytes = blockwait.to_bytes(length=32, byteorder='big')
        enc.bytes32(blockwait_bytes)
        enc.uint256(block_deadline)
        enc.uintn(target_vote_ppm, 24)
        data = add_0x(enc.get())
        tx = self.template(sender_address, contract_address, use_nonce=True)
        tx = self.set_code(tx, data)
        tx = self.finalize(tx, tx_format, id_generator=id_generator)
        return tx


    def add_option(self, contract_address, sender_address, proposal_idx, description, tx_format=TxFormat.JSONRPC, id_generator=None):
        enc = ABIContractEncoder()
        enc.method('addOption')
        enc.typ(ABIContractType.UINT256)
        enc.typ(ABIContractType.BYTES32)
        enc.uint256(proposal_idx)
        enc.bytes32(description)
        data = add_0x(enc.get())
        tx = self.template(sender_address, contract_address, use_nonce=True)
        tx = self.set_code(tx, data)
        tx = self.finalize(tx, tx_format, id_generator=id_generator)
        return tx


#    def propose(self, contract_address, sender_address, description, block_deadline, target_vote_ppm=500000, options=[], tx_format=TxFormat.JSONRPC, id_generator=None):
#        enc = ABIContractEncoder()
#        if len(options) == 0: 
#            enc.method('propose')
#        else:
#            enc.method('proposeMulti')
#        enc.typ(ABIContractType.BYTES32)
#        if len(options) > 0: 
#            enc.typ_literal('bytes32[]')
#        enc.typ(ABIContractType.UINT256)
#        enc.typ_literal('uint24')
#        enc.bytes32(description)
#        if len(options) > 0: 
#            enc.uint256(32*4)
#        enc.uint256(block_deadline)
#        enc.uintn(target_vote_ppm, 24)
#        if len(options) > 0: 
#            enc.uint256(len(options))
#            for v in options:
#                enc.bytes32(v)
#        data = add_0x(enc.get())
#        tx = self.template(sender_address, contract_address, use_nonce=True)
#        tx = self.set_code(tx, data)
#        tx = self.finalize(tx, tx_format, id_generator=id_generator)
#        return tx


    def vote(self, contract_address, sender_address, value, option=None, tx_format=TxFormat.JSONRPC, id_generator=None):
        enc = ABIContractEncoder()
        if option == None:
            enc.method('vote')
            enc.typ(ABIContractType.UINT256)
        else:
            enc.method('voteOption')
            enc.typ(ABIContractType.UINT256)
            enc.typ(ABIContractType.UINT256)
        if option != None:
            enc.uint256(option)
        enc.uint256(value)
        data = add_0x(enc.get())
        tx = self.template(sender_address, contract_address, use_nonce=True)
        tx = self.set_code(tx, data)
        tx = self.finalize(tx, tx_format, id_generator=id_generator)
        return tx


    def vote_cancel(self, contract_address, sender_address, value, tx_format=TxFormat.JSONRPC, id_generator=None):
        enc = ABIContractEncoder()
        enc.method('voteCancel')
        enc.typ(ABIContractType.UINT256)
        enc.uint256(value)
        data = add_0x(enc.get())
        tx = self.template(sender_address, contract_address, use_nonce=True)
        tx = self.set_code(tx, data)
        tx = self.finalize(tx, tx_format, id_generator=id_generator)
        return tx


    def scan(self, contract_address, sender_address, proposal_index, count, tx_format=TxFormat.JSONRPC, id_generator=None):
        enc = ABIContractEncoder()
        enc.method('scan')
        enc.typ(ABIContractType.UINT256)
        enc.typ(ABIContractType.UINT8)
        enc.uint256(proposal_index)
        enc.uintn(count, 8)
        data = add_0x(enc.get())
        tx = self.template(sender_address, contract_address, use_nonce=True)
        tx = self.set_code(tx, data)
        tx = self.finalize(tx, tx_format, id_generator=id_generator)
        return tx


    def finalize_vote(self, contract_address, sender_address, tx_format=TxFormat.JSONRPC, id_generator=None):
        enc = ABIContractEncoder()
        enc.method('finalize')
        data = add_0x(enc.get())
        tx = self.template(sender_address, contract_address, use_nonce=True)
        tx = self.set_code(tx, data)
        tx = self.finalize(tx, tx_format, id_generator=id_generator)
        return tx


    def withdraw(self, contract_address, sender_address, tx_format=TxFormat.JSONRPC, id_generator=None):
        enc = ABIContractEncoder()
        enc.method('withdraw')
        data = add_0x(enc.get())
        tx = self.template(sender_address, contract_address, use_nonce=True)
        tx = self.set_code(tx, data)
        tx = self.finalize(tx, tx_format, id_generator=id_generator)
        return tx


    def get_proposal(self, contract_address, proposal_idx, sender_address=ZERO_ADDRESS, id_generator=None):
        j = JSONRPCRequest(id_generator)
        o = j.template()
        o['method'] = 'eth_call'
        enc = ABIContractEncoder()
        enc.method('getProposal')
        enc.typ(ABIContractType.UINT256)
        enc.uint256(proposal_idx)
        data = add_0x(enc.get())
        tx = self.template(sender_address, contract_address)
        tx = self.set_code(tx, data)
        o['params'].append(self.normalize(tx))
        o['params'].append('latest')
        o = j.finalize(o)
        return o


    def get_option(self, contract_address, proposal_idx, option_idx, sender_address=ZERO_ADDRESS, id_generator=None):
        j = JSONRPCRequest(id_generator)
        o = j.template()
        o['method'] = 'eth_call'
        enc = ABIContractEncoder()
        enc.method('getOption')
        enc.typ(ABIContractType.UINT256)
        enc.typ(ABIContractType.UINT256)
        enc.uint256(proposal_idx)
        enc.uint256(option_idx)
        data = add_0x(enc.get())
        tx = self.template(sender_address, contract_address)
        tx = self.set_code(tx, data)
        o['params'].append(self.normalize(tx))
        o['params'].append('latest')
        o = j.finalize(o)
        return o


    def option_count(self, contract_address, proposal_idx, sender_address=ZERO_ADDRESS, id_generator=None):
        j = JSONRPCRequest(id_generator)
        o = j.template()
        o['method'] = 'eth_call'
        enc = ABIContractEncoder()
        enc.method('optionCount')
        enc.typ(ABIContractType.UINT256)
        enc.uint256(proposal_idx)
        data = add_0x(enc.get())
        tx = self.template(sender_address, contract_address)
        tx = self.set_code(tx, data)
        o['params'].append(self.normalize(tx))
        o['params'].append('latest')
        o = j.finalize(o)
        return o


    def vote_count(self, contract_address, proposal_idx, option_idx=0, sender_address=ZERO_ADDRESS, id_generator=None):
        j = JSONRPCRequest(id_generator)
        o = j.template()
        o['method'] = 'eth_call'
        enc = ABIContractEncoder()
        enc.method('voteCount')
        enc.typ(ABIContractType.UINT256)
        enc.typ(ABIContractType.UINT256)
        enc.uint256(proposal_idx)
        enc.uint256(option_idx)
        data = add_0x(enc.get())
        tx = self.template(sender_address, contract_address)
        tx = self.set_code(tx, data)
        o['params'].append(self.normalize(tx))
        o['params'].append('latest')
        o = j.finalize(o)
        return o


    def block_wait_limit(self, contract_address, sender_address=ZERO_ADDRESS, id_generator=None):
        j = JSONRPCRequest(id_generator)
        o = j.template()
        o['method'] = 'eth_call'
        enc = ABIContractEncoder()
        enc.method('blockWaitLimit')
        data = add_0x(enc.get())
        tx = self.template(sender_address, contract_address)
        tx = self.set_code(tx, data)
        o['params'].append(self.normalize(tx))
        o['params'].append('latest')
        o = j.finalize(o)
        return o


    def current_proposal(self, contract_address, sender_address=ZERO_ADDRESS, id_generator=None):
        j = JSONRPCRequest(id_generator)
        o = j.template()
        o['method'] = 'eth_call'
        enc = ABIContractEncoder()
        enc.method('getCurrentProposal')
        data = add_0x(enc.get())
        tx = self.template(sender_address, contract_address)
        tx = self.set_code(tx, data)
        o['params'].append(self.normalize(tx))
        o['params'].append('latest')
        o = j.finalize(o)
        return o



    @classmethod
    def parse_proposal(self, v, serial=None):
        v = strip_0x(v)
        logg.debug("proposal {}".format(v))

        cursor = 64
        dec = ABIContractDecoder()
        dec.typ(ABIContractType.BYTES32)
        dec.typ(ABIContractType.UINT256)
        dec.typ(ABIContractType.UINT256)
        dec.typ(ABIContractType.UINT256)
        dec.typ(ABIContractType.UINT256)
        dec.typ(ABIContractType.UINT256) # actually uint24
        dec.typ(ABIContractType.ADDRESS)
        dec.typ(ABIContractType.UINT8)

        dec.val(v[cursor:cursor+64]) # description
        cursor += 64 # options pos
        cursor += 64 # optionsvotes pos
        cursor += 64
        dec.val(v[cursor:cursor+64])
        cursor += 64
        dec.val(v[cursor:cursor+64])
        cursor += 64
        dec.val(v[cursor:cursor+64])
        cursor += 64
        dec.val(v[cursor:cursor+64])
        cursor += 64
        dec.val(v[cursor:cursor+64])
        cursor += 64
        dec.val(v[cursor:cursor+64])
        cursor += 64
        dec.val(v[cursor:cursor+64])
        cursor += 64

        r = dec.get()
        o = Proposal(r[0],
                     cancelVotes=r[1],
                     supply=r[2],
                     total=r[3],
                     block_deadline=r[4],
                     target_vote_ppm=r[5],
                     proposer=r[6],
                     state=r[7],
                     serial=serial,
                     )
        return o


def bytecode(**kwargs):
    return Voter.bytecode(version=kwargs.get('version'))


def create(**kwargs):
    enc = CLIEncoder()
    (typ, token_address) = enc.translate('a', strip_0x(kwargs['token_address']))
    voter_registry = kwargs.get('voter_registry')
    if voter_registry != None:
        (typ, voter_registry) = enc.translate('a', strip_0x(voter_registry))
    proposer_registry = kwargs.get('proposer_registry')
    if proposer_registry != None:
        (typ, proposer_registry) = enc.translate('a', strip_0x(proposer_registry))
    return Voter.cargs(token_address=token_address, protect_supply=kwargs.get('protect_supply'), voter_registry=voter_registry, proposer_registry=proposer_registry, version=kwargs.get('version'))


def args(v):
    if v == 'create':
        return (['token_address'], ['protect_supply', 'voter_registry', 'propose_registry'],)
    elif v == 'default' or v == 'bytecode':
        return ([], ['version'],)
    raise ValueError('unknown command: ' + v)
